test/e2e/xfn/xfn_test.go

```diff
@@ -0,0 +1,297 @@
+//go:build integration
+// +build integration
+
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package xfn
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"flag"
+	"fmt"
+	"github.com/crossplane/crossplane/apis/apiextensions/fn/proto/v1alpha1"
+	"google.golang.org/protobuf/proto"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"math/big"
+	"net"
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"strings"
+	"testing"
+	"time"
+)
+
+var xfnImageRef *string = flag.String("xfn-image-ref", "", "xfn runner image ref")
+
+func TestCustomCA(t *testing.T) {
+	cases := map[string]struct {
+		body func(xfnImageRef, certDir string) error
+	}{
+		"retrieve xfn image from private registry by setting CA cert via CLI arg": {
+			body: func(fnImageRef, certDir string) error {
+				// start xfn runner container
+				cmd := exec.Command("docker",
```

Comment:
I would rather avoid relying so much on docker and running pods, isn't there any other way to test this? Otherwise I'd rather integrate it with the e2es as many test case are going to pull stuff from a registry, which is now a public registry with some credentials, so it might be worth solving it once and for all there, maybe spinning up a registry inside or outside the cluster and putting in place the necessary networking to be able to pull stuff from there.

Reply:
Agree with @phisco here. Too many command invocations, likely to cause potential headaches.

Reply:
well, we need to spin `crossplane-xfn` container somewhere, and try to pull images from various registries. Ofcourse, we could also run a real e2e test, but then the fixture would be even larger - we would need to create composition, function and put them in some registry (with custom certificate) in order to verify that we are able to pull using custom CA bundle.

This is an integration test, does not require kubernetes to run, does not require crossplane to be installed, and we can in an easy way test some aspects of xfn runner. And docker is a dependency on any machine where we need to build and test.

Please let me know if we should drop these integration tests and go with full fledged e2e tests.

Reply:
> Please let me know if we should drop these integration tests and go with full fledged e2e tests.

Yes please - stick to the newly introduced E2E test pattern for now. If we feel there's a need to also have integration tests in addition to E2E and unit tests I'd like to discuss that first rather than organically introduce them alongside as part of feature PR.